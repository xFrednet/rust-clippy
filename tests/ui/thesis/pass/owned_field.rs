//@rustc-env: CLIPPY_PETS_PRINT=1
//@rustc-env: CLIPPY_PRINT_MIR=1

#[derive(Default)]
struct Example {
    owned_1: String,
    owned_2: String,
    copy_1: u32,
    copy_2: u32,
}

#[warn(clippy::borrow_pats)]
fn replace_self() {
    let mut drop = Example::default();
    drop = Example::default();

    let mut copy = 15;
    copy = 17;
}

#[warn(clippy::borrow_pats)]
fn conditional_replace_self() {
    let mut drop = Example::default();
    if true {
        drop = Example::default();
    }

    let mut copy = 15;
    if true {
        copy = 17;
    }
}

#[warn(clippy::borrow_pats)]
fn assign_copy_field() {
    let mut ex = Example::default();
    ex.copy_1 = 10;
}

#[warn(clippy::borrow_pats)]
fn assign_drop_field() {
    let mut ex = Example::default();
    ex.owned_1 = String::new();
}

#[warn(clippy::borrow_pats)]
fn move_drop_field_to_var() {
    let ex = Example::default();
    let _hey = ex.owned_1;
}

#[warn(clippy::borrow_pats)]
fn copy_field() {
    let ex = Example::default();
    let _hey = ex.copy_1;
}

#[warn(clippy::borrow_pats)]
fn move_drop_field_as_arg() {
    let ex = Example::default();
    take_string(ex.owned_1);
}

#[warn(clippy::borrow_pats)]
fn return_drop_field() -> String {
    let ex = Example::default();
    ex.owned_1
}

#[warn(clippy::borrow_pats)]
fn borrow_field_as_arg() {
    let ex = Example::default();
    take_string_ref(&ex.owned_1);
}

#[warn(clippy::borrow_pats)]
fn borrow_field_into_var() {
    let ex = Example::default();
    let _hey = &ex.owned_1;
}

fn take_string(_: String) {}
fn take_string_ref(_: &String) {}

fn main() {}
