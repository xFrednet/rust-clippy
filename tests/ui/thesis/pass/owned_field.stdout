# "replace_self"
- drop           : (Mutable  , Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Overwrite, DropForReplacement}
- copy           : (Mutable  , Owned    , Local   , Copy   , NonDrop , Primitive) {Overwrite}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments}

# "conditional_replace_self"
- drop           : (Mutable  , Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Overwrite, ConditionalOverwride, ConditionalDrop, DropForReplacement}
- copy           : (Mutable  , Owned    , Local   , Copy   , NonDrop , Primitive) {Overwrite, ConditionalOverwride}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments}

# "assign_copy_field"
- ex             : (Mutable  , Owned    , Local   , NonCopy, PartDrop, UserDef  ) {OverwritePart}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments}

# "assign_drop_field"
- ex             : (Mutable  , Owned    , Local   , NonCopy, PartDrop, UserDef  ) {OverwritePart}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments}

# "move_drop_field_to_var"
- ex             : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {PartMoved, PartMovedToVar}
- _hey           : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments}

# "copy_field"
- ex             : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {CopiedToVar}
- _hey           : (Immutable, Owned    , Local   , Copy   , NonDrop , Primitive) {}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments}

# "move_drop_field_as_arg"
- ex             : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {PartMoved, PartMovedToFn}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments}

# "return_drop_field"
- ex             : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {PartMoved, PartMovedToReturn}
- Body: Output(UserDef)  , NotConst , Safe  , Sync , Free       {NoArguments}

# "borrow_field_as_arg"
- ex             : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {PartBorrow, PartArgBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments, HasTempBorrow}

# "borrow_field_into_var"
- ex             : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {PartBorrow, PartNamedBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments, HasNamedBorrow}

